lib/bridge/MatrixHandler.js

```diff
@@ -1261,7 +1263,52 @@ MatrixHandler.prototype._sendIrcAction = Promise.coroutine(
         return;
     }
 
+    /**
+     * Cache events in here so we can refer to them for replies.
+     * We deliberately avoid fetching events from the homeserver due to speed concerns.
+     */
+    this._eventCache.set(event.event_id, { body: event.content.body, sender: event.sender });
+    if (this._eventCache.size > EVENT_CACHE_SIZE) {
+        const delKey = this._eventCache.entries().next().value[0];
+        this._eventCache.delete(delKey);
+    }
+
     let text = event.content.body;
+    const REPLY_SOURCE_MAX_LENGTH = 32;
+    const REPLY_NAME_MAX_LENGTH = 12;
+    if (event.content["m.relates_to"] && event.content["m.relates_to"]["m.in_reply_to"]) {
+        const eventId = event.content["m.relates_to"]["m.in_reply_to"].event_id;
+        // This is a reply.
+        const match = /> <(@.*:.*)>(.*)\n\n(.*)/.exec(text);
+        if (match.length === 4) {
+            let rplName = match[1];
+            let rplSource = match[2].trim();
+            if (this._eventCache.has(eventId)) {
+                rplName = this._eventCache.get(eventId).sender;
+                rplSource = this._eventCache.get(eventId).body;
+            }
+            // Replace newlines so we can fit this in the message body.
+            // It's not a perfect solution but it's legible.
+            rplSource = rplSource.replace(/\v/, '. ');
```

Comment:
or just grab the first line and call it a day. Modifying user's text like this feels too invasive for a bridge.

Reply:
Grabbing the first line makes sense...until someone does some really strange paragraphing. I'm not overly worried about editing out some newlines given the original text will be in IRC scrollback, and this version isn't too far removed from the original.

Reply:
This is going to be the thing I'm going to push back on fairly strongly: where possible we shouldn't be modifying a user's message - that's not the job of the bridge. This includes format markers which may introduce less clarity.

If someone does something weird with paragraphs, that's their own fault in my honest opinion. For a safety net, we can grab the first non-empty line of text.